HW4.py

```diff
@@ -0,0 +1,316 @@
+    p_max = max(predictor_list)
+    p_range = p_max - p_min
+    n_of_bin = 10
+    bin_width = p_range / n_of_bin
+
+    bin_list = [p_min]
+    s = 0
+    while s < p_max:
+        s += bin_width
+        bin_list.append(s)
+
+    df_bin = pd.DataFrame({"predictor": pd.Series(predictor_list)})
+    df_bin["LowerBin_UpperBin"] = pd.cut(
+        x=pd.Series(predictor_list), bins=bin_list, include_lowest=True
+    )
+    bincenter = []
+    for bin_n in df_bin["LowerBin_UpperBin"]:
+        bincenter.append(bin_n.mid)
+    df_bincenters = pd.DataFrame({"BinCenters": pd.Series(bincenter)})
+    df_bin["BinCenters"] = df_bincenters
+    df_bin["response"] = pd.Series(np.repeat(y, count_column - 1))
+
+    # Groupby df_bin table to create a Difference with mean table
+
+    df_bin_groupby = df_bin.groupby(("LowerBin_UpperBin"), as_index=False).agg(
+        bin_mean=pd.NamedAgg(column="predictor", aggfunc="mean"),
+        bin_count=pd.NamedAgg(column="predictor", aggfunc="count"),
+    )
+
+    bin_center_list = []
+    for bin_center in df_bin_groupby["LowerBin_UpperBin"]:
+        bin_center_list.append(bin_center.mid)
+
+    df_bin_groupby["BinCenter"] = pd.Series(bin_center_list)
+
+    PopulationMean = (np.sum(X) + np.sum(y)) / (count_row * count_column)
+    df_bin["PopulationMean"] = PopulationMean
+    df_bin_groupby["PopulationMean"] = PopulationMean
+
+    MeanSquaredDiff = (
+        df_bin_groupby["bin_mean"] - df_bin_groupby["PopulationMean"]
+    ) ** 2
+    df_bin_groupby["MeanSquaredDiff"] = MeanSquaredDiff
+
+    print(df_bin_groupby)
+
+    # Square the difference, sum them up and divide by number of bins
+    print(
+        f"THE unWeighted NUMBER IS : {df_bin_groupby['MeanSquaredDiff'].sum() / n_of_bin}"  # noqa
+    )
+
+    fig = px.bar(df_bin_groupby, x="BinCenter", y="bin_count")
+    fig.show()
+
+    # Difference with mean table (weighted)
+    print("***Difference with mean table (weighted)***")
+
+    df_bin_groupby_weighted = df_bin_groupby.copy()
+
+    population_proportion = []
+    for count in df_bin_groupby["bin_count"]:
+        population_proportion.append(count / len(predictor_list))
+
+    df_bin_groupby_weighted["PopulationProportion"] = pd.Series(
+        population_proportion
+    )  # noqa
+    df_bin_groupby_weighted["MeanSquaredDiffWeighted"] = (
+        df_bin_groupby_weighted["MeanSquaredDiff"]
+        * df_bin_groupby_weighted["PopulationProportion"]
+    )
+
+    # Square the difference, sum them up and divide by number of bins
+    print(
+        f"THE Weighted NUMBER IS :\
+        {df_bin_groupby_weighted['MeanSquaredDiffWeighted'].sum() / n_of_bin}"
+    )
+
+    print(df_bin_groupby_weighted)
+
+    fig = px.bar(df_bin_groupby_weighted, x="BinCenter", y="bin_count")
+    fig.show()
+
+    # Random Forest Variable importance ranking
+    print("***Random Forest Variable importance ranking***")
+
+    X_train, X_test, y_train, y_test = train_test_split(X, y)

```

Comment:
Don't worry about running a train_test_split for the feature importances

Reply:
updated

Reply:
Nice catch!